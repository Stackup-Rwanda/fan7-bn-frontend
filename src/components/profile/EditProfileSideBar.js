import React, { Component } from 'react';
import PropTypes from 'prop-types';
import swal from 'sweetalert';
import Avatar from './ProfileAvatar';
import './styles/editProfileSideBar.scss';
import defaultImg from '../../assets/images/download (1).jpeg';

class EditProfileSideBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      role: '',
      profileImg: null,
    };

    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    const { email, role, profileImg } = this.props;
    this.setState({
      email,
      role,
      profileImg,
    });
  }

  handleChange(event) {
    event.persist();
    const { handleChange } = this.props;
    const imageFile = event.target.files[0];
    if (!imageFile) return;
    if (!imageFile.name.match(/\.(jpg|jpeg|png)$/)) {
      swal({
        title: 'Error',
        text: 'Please select valid image.',
        icon: 'error',
        buttons: 'Close',
      });
      return;
    }
    this.setState(() => ({
      profileImg: URL.createObjectURL(imageFile),
    }));
    handleChange(imageFile);
  }

  render() {
    const { email, role, profileImg } = this.state;

    return (
      <div className="sideBar">
        <div className="sideBar__avatar">
          <Avatar src={profileImg || defaultImg} />
        </div>
        <label className="sideBar__change" htmlFor="input-file">
          change profile
          <input
            type="file"
            id="input-file"
            className="sideBar__change___file"
            onChange={this.handleChange}
          />
        </label>
        <p className="sideBar__text">
          <strong>Email: </strong>
          {email}
        </p>
        <p className="sideBar__text">
          <strong>Role: </strong>
          {role}
        </p>
      </div>
    );
  }
}

EditProfileSideBar.propTypes = {
  email: PropTypes.string,
  role: PropTypes.string,
  profileImg: PropTypes.string,
  handleChange: PropTypes.func.isRequired,
};

EditProfileSideBar.defaultProps = {
  email: '',
  role: '',
  profileImg: '',
};

export default EditProfileSideBar;
