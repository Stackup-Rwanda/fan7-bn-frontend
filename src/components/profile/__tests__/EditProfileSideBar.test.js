/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import EditProfileSideBar from '../EditProfileSideBar';

describe('EditProfileSideBar Component Test', () => {
  it('should return true if component exists', () => {
    const wrapper = mount(<EditProfileSideBar />);

    expect(wrapper.find('.sideBar').exists()).toBe(true);
  });

  it('should display appropriate props', () => {
    const props = {
      email: 'elvisrugamba@gmail.com',
    };
    const wrapper = mount(<EditProfileSideBar {...props} />);

    expect(
      wrapper
        .find('.sideBar__text')
        .first()
        .text()
    ).toEqual('Email: elvisrugamba@gmail.com');
  });

  it('should not handle change if image is not provided', () => {
    global.URL.createObjectURL = jest.fn();
    const handleChange = jest.fn();
    const event = {
      persist: jest.fn(),
      target: {
        files: [],
      },
    };
    const props = {
      handleChange,
    };
    const wrapper = shallow(<EditProfileSideBar {...props} />);

    wrapper.find('input[type="file"]').simulate('change', event);

    expect(handleChange).not.toHaveBeenCalled();
  });

  it('should not handle change if image is invalid', () => {
    global.URL.createObjectURL = jest.fn();
    const handleChange = jest.fn();
    const event = {
      persist: jest.fn(),
      target: {
        files: [{ name: 'image.txt' }],
      },
    };
    const props = {
      handleChange,
    };
    const wrapper = shallow(<EditProfileSideBar {...props} />);

    wrapper.find('input[type="file"]').simulate('change', event);

    expect(handleChange).not.toHaveBeenCalled();
  });

  it('should handle change if image is valid', () => {
    global.URL.createObjectURL = jest.fn();
    const handleChange = jest.fn();
    const event = {
      persist: jest.fn(),
      target: {
        files: [{ name: 'image.jpg' }],
      },
    };
    const props = {
      handleChange,
    };
    const wrapper = shallow(<EditProfileSideBar {...props} />);

    wrapper.find('input[type="file"]').simulate('change', event);

    expect(handleChange).toHaveBeenCalled();
  });
});

describe('<EditProfileSideBar /> shallow rendering tests', () => {
  it('matches the snapshot', () => {
    const props = {
      email: 'elvisrugamba@gmail.com',
      role: 'manager',
      handleChange: jest.fn(),
    };
    const tree = shallow(<EditProfileSideBar {...props} />);

    expect(tree).toMatchSnapshot();
  });
});
