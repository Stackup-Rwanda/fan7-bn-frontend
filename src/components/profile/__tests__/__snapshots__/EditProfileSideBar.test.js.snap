// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProfileSideBar /> shallow rendering tests matches the snapshot 1`] = `
<div
  className="sideBar"
>
  <div
    className="sideBar__avatar"
  >
    <ProfileAvatar
      src={
        Object {
          "process": [Function],
        }
      }
    />
  </div>
  <label
    className="sideBar__change"
    htmlFor="input-file"
  >
    change profile
    <input
      className="sideBar__change___file"
      id="input-file"
      onChange={[Function]}
      type="file"
    />
  </label>
  <p
    className="sideBar__text"
  >
    <strong>
      Email: 
    </strong>
    elvisrugamba@gmail.com
  </p>
  <p
    className="sideBar__text"
  >
    <strong>
      Role: 
    </strong>
    manager
  </p>
</div>
`;
